### DISCUSSION TITLE
Kanidm behind Caddy reverse proxy using docker compose

### DISCUSSION THREAD

**Discussion by ksmolder** [kanidm/kanidm#3322] [Q&A] [answered]
# What I'm trying to accomplish

Setting up kanidm behind a Caddy reverse proxy. Both Caddy and kanidm are deployed as docker containers through docker compose. A dedicated subdomain (_idm_) is used of a domain that I "own" (referred to as _[domain.com]_).

# Initial setup and configuration files
**> Caddyfile**
``` Caddyfile
idm.[domain.com]:443 {
  reverse_proxy https://kanidm:8443
  log {
    output stdout
  }
}
```

**> Kanidm server.toml**
``` server.toml
bindaddress = "[::]:8443"
db_path = "/data/kanidm.db"
tls_chain = "/data/idm.[domain.com].crt"
tls_key = "/data/idm.[domain.com].key"
domain = "idm.[domain.com]"
origin = "https://idm.[domain.com]"

[online_backup]
path = "/data/backups/"
schedule = "00 22 * * *"
```

**Note**: the certificates are copies of the certificates managed by Caddy (signed by Let's Encrypt). 

**> docker-compose.yml**
``` docker-compose.yml
services:

  caddy:
    container_name: caddy
    image: caddy:2.8
    restart: unless-stopped
    environment:
      - TZ=Europe/Brussels
    ports:
      - 80:80
      - 443:443
      - 443:443/udp
    user: 33:33 # www-data:www-data
    volumes:
      - /srv/docker/caddy/etc/caddy:/etc/caddy
      - /srv/docker/caddy/data:/data
      - /srv/docker/caddy/config:/config
      - /srv/www:/srv:ro


  kanidm:
    container_name: kanidm
    image: kanidm/server:1.4.5
    restart: unless-stopped
    expose:
      - 8443
    volumes:
      - /srv/docker/kanidm/data:/data
```

# Observed behaviour and workaround

Whenever I navigate to https://idm.[domain.com], Caddy returns an HTTP response code 502. Caddy logs the request as error. Kanidm does not log anything.

What I've discovered is that once I change the `container_name` property of the kanidm container to _idm.[domain.com]_ and update the Caddy file to proxy the request to _https://idm.[domain.com]:8443_, everything seems to work.

**> Caddyfile**
``` Caddyfile
idm.[domain.com]:443 {
  reverse_proxy https://idm.[domain.com]:8443
  log {
    output stdout
  }
}
```
However, I'd like to avoid having to use my public DNS names in my docker compose files as the actual names of a container.

# What I've already tried

In an attemtpt to get things working using _kanidm_ as the `container_name` for kanidm, I've tried to add some configuration to Caddy to rewrite some HTTP headers (experimented with different combinations of the `Host` and `X-Forwarded-Host` + `X-Forwarded-Proto` headers):

**> Caddyfile**
``` Caddyfile
idm.[domain.com]:443 {
  reverse_proxy https://kanidm:8443 {
    header_up Host "idm.[domain.com]:8443"
#    header_up X-Forwarded-Host "idm.[domain.com]:443"
#    header_up X-Forwarded-Proto "https"
  }
  log {
    output stdout
  }
}
```

However, this does not solve the issue.

# Specific question

Are there any suggestions or examples of how to setup kanidm behind Caddy? When using Traefik, it [seems possible](https://kanidm.github.io/kanidm/stable/examples/traefik.html) to stick to an internal container name like _kanidm_. So I assume that with the correct reverse proxy configuration, the same should be possibe for Caddy...

Thanks for any suggestions!

**Comment by yaleman** [comment, marked as answer]
I think you should be able to set the [tls_server_name](https://caddyserver.com/docs/caddyfile/directives/reverse_proxy#tls_server_name) to the hostname so caddy knows what to expect on the cert.

**Comment by qiuwei** [comment]
I have a follow-up question.
How can I automatically sync the certificates from caddy to kanidm containers?
The caddy container renews the certificates at regular intervals. Can the certificates used by Kanidm be updated without any downtime?

**Reply by Kasi07** [reply]
According to the documentation:
```
#   TLS chain and key in pem format. Both must be present.
#   If the server receives a SIGHUP, these files will be
#   re-read and reloaded if their content is valid.
```

I use acme.sh and have a separate docker container which just listens for a post request and then executes `kill -HUP 1` in the pid space of the server.

`docker-compose.yml`:
```
  reload-tls-certs:
    build:
      context: ./reload-tls-certs
    cap_drop:
      - ALL
    cap_add:
      - KILL
    pid: service:kanidm_server
    depends_on:
      kanidm_server:
        condition: service_healthy
```

`./reload-tls-certs/Dockerfile`:
```
FROM busybox:latest
COPY server.sh /server.sh
RUN chmod +x /server.sh
EXPOSE 80
CMD ["/server.sh"]
```

`./reload-tls-certs/server.sh`:
```
#!/bin/sh
PORT=80
COMMAND="kill -HUP 1"

while true; do
  {
    # read & discard HTTP request headers
    while IFS= read -r line && [ "$line" != "" ]; do :; done

    # send minimal HTTP response headers
    printf 'HTTP/1.1 200 OK\r\n'
    printf 'Content-Type: text/plain; charset=utf-8\r\n'
    printf '\r\n'

    # run your command
    $COMMAND
  } | nc -l -p $PORT
done
```

And using acme.sh I then set a post request to this container as the renew-hook.

**Reply by andreas-bulling** [reply]
@Kasi07 Thanks for the solution. I tried to make it work on my setup but while the busybox container starts up properly - and `docker ps` shows that it has port 80 open - I don't see any process listening on port 80. Is it because this is a privileged port?

And how does your POST renew-hook look like? Something like `curl -X POST -d 'test' localhost:80`?

EDIT:

This is needed to make the port available on the docker host:

```
ports:
      - "80:80"
```

**Comment by yaleman** [comment]
Another example here: https://github.com/kanidm/kanidm/discussions/3777
